impl/src/main/java/impl/UserResourceImpl.java

```diff
@@ -71,16 +80,47 @@ public UserResourceImpl(UserDao userDao, ResponseHeaderHolder responseHeaderHold
     public Observable<User> generateToken(@NotNull String openIdToken) {
         return openIdValidator.validate(openIdToken).flatMap(validOpenId ->
                 userDao.getUserByEmail(validOpenId.email)
-                .onErrorResumeNext(t -> error(new WebException(HttpResponseStatus.INTERNAL_SERVER_ERROR, "failed.to.search.for.user", t)))
+                .onErrorResumeNext(t -> error(new WebException(HttpResponseStatus.INTERNAL_SERVER_ERROR, FAILED_TO_SEARCH_FOR_USER, t)))
                 .switchIfEmpty(addUserToDatabase(validOpenId.name, validOpenId.email, validOpenId.picture))
-                .map(user -> {
-                            ApplicationToken applicationToken = applicationTokenCreator.createApplicationToken(validOpenId, user.getId());
-                            addAsCookie(applicationToken, user);
-                            return user;
-                        }
-                ));
+                .flatMap(user -> updateUserNameAndPicture(user, validOpenId))
+                .map(user -> addApplicationTokenToHeader(validOpenId, user)));
+    }
+
+    private User addApplicationTokenToHeader(ImmutableOpenIdToken validOpenId, User user) {
+        ApplicationToken applicationToken = applicationTokenCreator.createApplicationToken(validOpenId, user.getId());
+        addAsCookie(applicationToken, user);
+        return user;
     }
 
+    private Observable<User> updateUserNameAndPicture(User user, ImmutableOpenIdToken openId) {
+        if(!hasUpdatedNameOrPicture(user, openId)) {
+            return just(user);
+        }
+
+        return userDao.updateUser(user.getId(), openId.name, openId.picture)
+            .doOnError(e -> LOG.error("failed to update user with id: {} with the following update name: {} and updated picture: {}",
+                user.getId(),
+                openId.name,
+                openId.picture))
+            .onErrorResumeNext(e -> error(new WebException(HttpResponseStatus.INTERNAL_SERVER_ERROR, FAILED_TO_UPDATE_USER_NAME_OR_PICTURE, e)))
+            .map(e -> {
+                user.setName(openId.name);
+                user.setPicture(openId.picture);
+                return user;
+            });
+    }
+
+    private boolean hasUpdatedNameOrPicture(User user, ImmutableOpenIdToken validOpenId) {
```

Comment:
Could be simplified to
return !Objects.equals(user.getName(), validOpenId.name) || !Objects.equals(user.getPicture(), validOpenId.picture)

Reply:
Yes , but i think that this inlines better with our code standards. @splitfeed ?

Reply:
Agree with @jepp3, guard clauses is preferred

Reply:
Yeah, guard clauses please :)